File: src/q3.py
import math

def find_largest_prime_factor():
    ques = 600851475143
    while True:
        smallest_factor = smallest_pf(ques)
        if smallest_factor < ques:
            ques //= smallest_factor
        else:
            return str(ques)


def smallest_pf(n):
    if n < 2:
        return None
    for divisor in range(2, math.isqrt(n) + 1):
        if n % divisor == 0:
            return divisor
    return n  

print(find_largest_prime_factor())

File: src/R/q3.R
smallest_prime_factor <- function(n) {
  for (i in 2:sqrt(n)) {
    if (n %% i == 0) {
      return(i)
    }
  }
  return(n)
}

find_largest_prime_factor <- function() {
  ques <- 600851475143
  while (TRUE) {
    smallest_factor <- smallest_prime_factor(ques)
    if (smallest_factor < ques) {
      ques <- ques / smallest_factor
    } else {
      return(ques)
    }
  }
}

print(find_largest_prime_factor())

